pkg/kn/commands/plugin/plugin_list.go

```diff
@@ -0,0 +1,190 @@
+// Copyright © 2019 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/knative/client/pkg/kn/commands"
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+// ValidPluginFilenamePrefixes controls the prefix for all kn plugins
+var ValidPluginFilenamePrefixes = []string{"kn"}
+
+// NewPluginListCommand creates a new `kn plugin list` command
+func NewPluginListCommand(p *commands.KnParams) *cobra.Command {
+	pluginFlags := PluginFlags{
+		IOStreams: genericclioptions.IOStreams{
+			In:     os.Stdin,
+			Out:    os.Stdout,
+			ErrOut: os.Stderr,
+		},
+	}
+
+	pluginListCommand := &cobra.Command{
+		Use:   "list",
+		Short: "List all visible plugin executables",
+		Long: `List all visible plugin executables.
+
+		Available plugin files are those that are:
+		- executable
+		- begin with "kn-
+		- anywhere on the path specfied in Kn's config pluginDir variable, which:
+		  * defaults to $PATH if not specified
+		  * can be overridden with the --plugin-dir flag`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := pluginFlags.complete(cmd)
+			if err != nil {
+				return err
+			}
+
+			err = pluginFlags.run()
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	pluginFlags.AddPluginFlags(pluginListCommand)
+
+	return pluginListCommand
+}
+
+// Private
+
+func (o *PluginFlags) complete(cmd *cobra.Command) error {
+	o.Verifier = &CommandOverrideVerifier{
+		Root:        cmd.Root(),
+		SeenPlugins: make(map[string]string, 0),
+	}
+
+	pluginPath := commands.PluginDir
+	if pluginPath == "$PATH" {
+		pluginPath = os.Getenv("PATH")
+	}
+
+	o.PluginPaths = filepath.SplitList(pluginPath)
+
+	return nil
+}
+
+func (o *PluginFlags) run() error {
+	pluginsFound := false
+	isFirstFile := true
+	pluginErrors := []error{}
+	pluginWarnings := 0
+
+	for _, dir := range uniquePathsList(o.PluginPaths) {
+		if dir == "" {
+			continue
+		}
+		files, err := ioutil.ReadDir(dir)
+		if err != nil {
```

Comment:
I would silently ignore any errors. Right now when I run it w/o any plugins I see this:
```
$ ./kn plugin list
Unable read directory "/root/src/k8s.io/kubernetes/_output/dockerized/bin/linux/amd64" from your pluginDir: open /root/src/k8s.io/kubernetes/_output/dockerized/bin/linux/amd64: no such file or directory. Skipping...
warning: unable to find any kn plugins in your pluginDir
```
I would expect no output instead. The error message does not help me at all and `kn` has not reason to care about dirs that I don't have access to but are on my PATH.
No plugins isn't an error or a warning-able thing.

Reply:
dude, it tells you that your pluginDir is not readable. I am happy to make the message clearer but I think silent ignore is not the correct approach.

Reply:
look at my example, I'm going to get that message every single time for no good reason. That's not a good UX.